### ISSUE TITLE
CRI: upgrade protobuf to v3

### ISSUE THREAD

**Issue by feiskyer** [kubernetes/kubernetes#38854]
We use proto2 in CRI because protobuf v3 is beta then. Now that protobuf [v3.1.0](https://github.com/google/protobuf/releases) have released for a few months, we should consider upgrade to v3 in case of API incompatibility in the future (see what's new in v3 at [here](https://github.com/google/protobuf/releases/tag/v3.0.0)).

According protocol buffer [developer guide](https://developers.google.com/protocol-buffers/docs/overview#introducing-proto3):

>  we recommend using proto3 for all new gRPC servers and clients as it avoids compatibility issues.

cc @yujuhong @kubernetes/sig-node-misc

**Comment by feiskyer** [comment]
Note that the biggest difference of proto3 is the [default values](https://developers.google.com/protocol-buffers/docs/proto3#default), there is no pointer to basic types any more (and no generated `GetXXX()` functions for them). The default values are type-specific:

1.  For strings, the default value is the empty string.
2.  For bytes, the default value is empty bytes.
3.  For bools, the default value is false.
4.  For numeric types, the default value is zero.
5.  For [enums](https://developers.google.com/protocol-buffers/docs/proto3#enum), the default value is the **first defined enum value**, which must be 0.
6.  For message fields, the field is not set. (nil in go)

Most of those changes work smoothly for CRI, except enums. We need to change the first item in enums to something unspecified (e.g. `ContainerStateNotSpecified`) for listing containers/sandboxes with all states. 

Added a WIP PR #39158 for testing proto v3.

cc @yujuhong @euank @mrunalp

**Comment by yujuhong** [comment]
@feiskyer what's the reason to upgrade to v3 so soon?

**Comment by feiskyer** [comment]
> what's the reason to upgrade to v3 so soon?

@yujuhong for a couple of reasons, it's better to upgrade to v3:

1. v2 and v3 are incompatible
2. v3 is recommended by protocol buffer: `we recommend using proto3 for all new gRPC servers and clients as it avoids compatibility issues.` (see [here](https://developers.google.com/protocol-buffers/docs/overview#introducing-proto3))
3. We use proto2 in CRI because protobuf v3 is beta then. Now protobuf v3.1.0 have released for a few months.

**Comment by Random-Liu** [comment]
@feiskyer Does protobuf v3 have better performance than v2? I've seen significant memory usage of grpc in my benchmark. Will post some result soon.

**Comment by feiskyer** [comment]
@Random-Liu Didn't see related reports. Could you help also make a benchmark of v3 based on #39158?

**Comment by yujuhong** [comment]
> v3 is recommended by protocol buffer: we recommend using proto3 for all new gRPC servers and clients as it avoids compatibility issues. (see here)

I also saw this in the [release note](https://github.com/google/protobuf/releases/tag/v3.0.0):
```
We recommend that new Protocol Buffers users use proto3. However, we do not
generally recommend that existing users migrate from proto2 from proto3 due
to API incompatibility, and we will continue to support proto2 for a long
time.
```

I don't think there is a strong incentive for us to switch to proto3. Compatibility wouldn't be an issue if everyone uses proto2. On the other hand, I do like the simpler proto3, and the generated go code (less pointers and functions).

This will affect all WIP runtimes. /cc @jonboulle @mrunalp

**Comment by yujuhong** [comment]
> Note that the biggest difference of proto3 is the default values, there is no pointer to basic types any more (and no generated GetXXX() functions for them). The default values are type-specific:
> 
> For strings, the default value is the empty string.
> For bytes, the default value is empty bytes.
> For bools, the default value is false.
> For numeric types, the default value is zero.
> For enums, the default value is the first defined enum value, which must be 0.
> For message fields, the field is not set. (nil in go)
> Most of those changes work smoothly for CRI, except enums. We need to change the first item in enums to something unspecified (e.g. ContainerStateNotSpecified) for listing containers/sandboxes with all states.
> 

The default values may cause problems for some integer fields. For example, `ExitCode` will be default to 0. We could use some message type:
https://developers.google.com/protocol-buffers/docs/reference/google.protobuf#google.protobuf.Int32Value

In addition to that, we test the presence of some fields in the api to determine what to do. The "filters" for listing the sandboxes/containers is one example. We should also consider wrapping individual fields in the `filter` struct.

**Comment by yujuhong** [comment]
@smarterclayton @lavalamp , do you have any suggestion about using proto3 vs 2?

**Comment by smarterclayton** [comment]
We're definitely not ready on the API side to move to 3 - we'd need to
change our representation to handle nil and empty fields, and some of the
support we wanted from gogo-protobuf weren't available in 3.  However, I
don't think there's any reason that CRI couldn't use 3 as long as we can
ensure that gogo-protobuf isn't broken when we update.

**Comment by yujuhong** [comment]
Upgrade to proto3:
Pros:
 - more readable go code (less dereferencing of pointers)
 - preliminary testing suggests lower memory usage than proto2

Cons:
 - proto2 and 3 are not compatible; this breaks all existing CRI implementation 
 - more churn before rollout

Thoughts? Objections?
/cc @kubernetes/sig-node-misc @johanatan @yifan-gu @mrunalp @sjpotter

**Comment by yifan-gu** [comment]
@yujuhong For cri I think we should upgrade to proto3 asap, we will have more to break if we are doing it later.
I can help on this.

**Comment by mrunalp** [comment]
@yujuhong I have no problems in migration cri-o to proto 3. Thanks!

**Comment by yujuhong** [comment]
> @yujuhong For cri I think we should upgrade to proto3 asap, we will have more to break if we are doing it later.
> I can help on this.

Thanks for offering! @feiskyer already has a PR for doing this: #39158, but it doesn't address the points in https://github.com/kubernetes/kubernetes/issues/38854#issuecomment-272045791. @yifan-gu do you want to help review the PR?

**Comment by yifan-gu** [comment]
@yujuhong Sure, will review in the afternoon.

**Comment by feiskyer** [comment]
> Thanks for offering! @feiskyer already has a PR for doing this: #39158, but it doesn't address the points in #38854 (comment).

Will update to wrapped types later today.

**Comment by Random-Liu** [comment]
> preliminary testing suggests lower memory usage than proto2

I'll run some more test against proto3, and post more result.

**Comment by yifan-gu** [comment]
Fixed by #39158 
It's so fast! Thanks for the work @feiskyer @yujuhong @Random-Liu !
